README.md
# REDCap Complex Field Validation

A REDCap Module that allows for the addition of valid options outside of the min-max range for a REDCap form field with number validation.


## Prerequisites
- [REDCap Modules](https://github.com/vanderbilt/redcap-external-modules)

## Installation
- Clone this repo into to `<redcap-root>/modules/complex_field_validation_v<version_number>`.
- Go to **Control Center > Manage External Modules** and enable Complex Field Validation.
- For each project you want to use this module, go to the project home page, click on **Manage External Modules** link, and then enable Complex Field Validation for that project.


## How to use
Once the module is activated on a project, you may include values outside the predefined min-max range either by 
1. adding a `<span>` tag with class `valid` on the *Field Note*, or by
2. adding the action tag `@EXTRA-VALID-RANGES` on the *Action Tags / Field Annotation*. 

For example, if the question were something like "In which month was the subject born?," simply use the 1-12 range and decorate additional valid options either in the 'Field Note' like this:

    1-12, or <span class="valid">88</span> if unknown.
    
or in the 'Action Tags / Field Annotation':

    @EXTRA-VALID-RANGES = "88"

You may include as many additional values as you need. As an illustration, the configuration below will allow the user to add the new valid values 88, 95, 96, and 97, although none of these values are in the preestablished range of 1-12.

    1-12, or (<span class="valid">88</span>,<span class="valid">95</span>, <span class="valid">96</span>, <span class="valid">97</span>)
    
To use the action tag instead (or in addition), you can include the following in the 'Action Tags / Field Annotation' field

    @EXTRA-VALID-RANGES = "88,95-97"

Notice that you can combine ranges and fixed values as such `@EXTRA-VALID-RANGES = "88,95-97,13,15-17,24"`.

These two ways of including values outside the predefined min-max range can be used simultaneously. 

CHANGELOG.md
# Change Log
All notable changes to the Complex Field Validation module will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2018-01-23
### Added
- Add @EXTRA-VALID-RANGE action tag that mimics functionality of the span in the field note (Marly)
- Add help in online designer for @EXTRA-VALID-RANGE action tag (Marly)

### Changed
- Improve code readability (Marly)
- Modify variable names to improve code readability and match code standards (Marly)


## [1.0] - 2018-01-02
### Summary
- This is the first release

### Added
- Revise description in config and README (Philip Chase)
- Set minimum redcap version to 8.0.3 to get support for semantic version numbers (Philip Chase)
- Add pbc to authors (Philip Chase)
- Revise description in config and README (Philip Chase)
- Merge pull request #1 from marlycormar/develop (Philip Chase)
- Add documentation (Marly Cormar)
- Add description (Marly Cormar)
- Include js file that does the logic behind module (Marly Cormar)
- Include js helper file (Marly Cormar)
- Add function that includes js files (Marly Cormar)
- Create skeleton of the config.json file (Marly Cormar)
- Create skeleton of the ExternalModules.php file (Marly Cormar)
- Initial commit (Philip Chase)



js/addTag.js
$(document).ready(function() {
    // Checking if field annotation is present on this page.
    if ($('#div_field_annotation').length === 0) {
        return false;
    }

    $('body').on('dialogopen', function(event, ui) {
        var $popup = $(event.target);
        if ($popup.prop('id') !== 'action_tag_explain_popup') {
            // That's not the popup we are looking for...
            return false;
        }

        // Aux function that checks if text matches the "@DEFAULT" string.
        var isDefaultLabelColumn = function() {
            return $(this).text() === '@DEFAULT';
        }

        // Getting @DEFAULT row from action tags help table.
        var $default_action_tag = $popup.find('td').filter(isDefaultLabelColumn).parent();
        if ($default_action_tag.length !== 1) {
            return false;
        }

        var tag_name = '@EXTRA-VALID-RANGES';
        var descr = 'TODO.';

        // Creating a new action tag row.
        var $new_action_tag = $default_action_tag.clone();
        var $cols = $new_action_tag.children('td');
        var $button = $cols.find('button');

        // Column 1: updating button behavior.
        $button.attr('onclick', $button.attr('onclick').replace('@DEFAULT', tag_name));

        // Columns 2: updating action tag label.
        $cols.filter(isDefaultLabelColumn).text(tag_name);

        // Column 3: updating action tag description.
        $cols.last().text(descr);

        // Placing new action tag.
        $new_action_tag.insertAfter($default_action_tag);
    });
});


js/checkRanges.js
$('document').ready(function() {
/*
 * Finds additional valid options outside of min-max ranges.
 *
 * Searches for field notes that have children elements of class "valid". For
 * all that are found, store the original REDCap validation (found in the
 * "onblur" event callback) and replace it with a custom one.
 *
 * If the user enters a value not found in the list of additional valid
 * options, the original REDCap validation is called.
 */

    function isValueOnRange(fieldName, val) {
      var ranges = actionTagValuesPerField[fieldName];

      if(ranges == undefined) return false;

      for(i = 0; i < ranges.length; i++){
        if(ranges[i].length == 1){
          if(val == ranges[i][0])
            return true;
        }
        if(ranges[i].length == 2){
          if(val >= ranges[i][0] && val <= ranges[i][1])
            return true;
        }
      }

      return false;
    }

    $( "input" ).filter(function( index, element ) {
      return $(element).siblings(".note:has('.valid')").length > 0 ||
      $(element).siblings("[name='age']").andSelf().length > 0;
    })
    .each(function( index, element ) {
      var additionalOptions = $(element).siblings(".note").find('.valid');
      var actionTagAdditionOptions = $(element).siblings().andSelf().find("[name='age']"); 
      var original = element.onblur;
      element.onblur = null;

      $(element).on( "blur", function( evt ) {
        var val = this.value;
        var matches = additionalOptions.filter(function( index, element ) {
          return $(element).text() == val;
        });

        if ( matches.length == 0  && isValueOnRange($(element).attr('name'),val) == 0) {
          $(this).trigger( "original", original );
        }
        else{
          element.style.backgroundColor="white";
        }
        
      }).on( "original", original );
    });
});




ExternalModule.php
<?php
/**
 * @file
 * Provides ExternalModule class for Complex Field Validation.
 */

namespace ComplexFieldValidation\ExternalModule;

use ExternalModules\AbstractExternalModule;
use ExternalModules\ExternalModules;
use Form;

/**
 * ExternalModule class for Complex Field Validation.
 */
class ExternalModule extends AbstractExternalModule {

    /**
     * @inheritdoc
     */
    function redcap_data_entry_form_top($project_id, $record, $instrument, $event_id, $group_id) {
        $this->includeJs('js/checkRanges.js');

        global $Proj;

        $action_tag_name = "@EXTRA-VALID-RANGES";
        // Array holding the (field name, action-tag values) pairs
        $action_tag_values_per_field = array();

        foreach (array_keys($Proj->forms[$_GET['page']]['fields']) as $field_name) {
            $field_info = $Proj->metadata[$field_name];

            if (!$display_mode = Form::getValueInActionTag($field_info['misc'], $action_tag_name)) {
                continue;
            }

            // Split the action tag values into an array of arrays
            $field_tag_values = explode(",", $display_mode);
            for($i = 0; $i < sizeof($field_tag_values); $i++){
                $field_tag_values[$i] = explode("-", $field_tag_values[$i]);
            }

            $action_tag_values_per_field[$field_name] =  $field_tag_values;
        }

        $this->sendVarToJS('actionTagValuesPerField', $action_tag_values_per_field);
    }


    /**
     * @inheritdoc
     */
    function redcap_every_page_top($project_id) {
        if (PAGE == 'Design/online_designer.php' && $project_id) {
            // Add action tag
            $this->includeJs('js/addTagToOnlineDesignerActionTagList.js');
        }
    }


    /**
     * Includes a local JS file.
     *
     * @param string $path
     *   The relative path to the js file.
     */
    protected function includeJs($path) {
        echo '<script src="' . $this->getUrl($path) . '"></script>';
    }


    /**
     * Sends a PHP variable over to JS.
     *
     * @param string $name
     *   Variable name
     * @param var $value
     *   Variable value
     */
    protected function sendVarToJS($name, $value) {
        echo '<script>var '. $name .' = ' . json_encode($value) . ';</script>';
    }
}


js/complexFieldValidationHelper.js
$('document').ready(function() {
/*
 * Finds additional valid options outside of min-max ranges.
 *
 * Searches for field notes that have children elements of class "valid". For
 * all that are found, store the original REDCap validation (found in the
 * "onblur" event callback) and replace it with a custom one.
 *
 * If the user enters a value not found in the list of additional valid
 * options, the original REDCap validation is called.
 */
    $( "input" ).filter(function( index, element ) {
      return $(element).siblings(".note:has('.valid')").length > 0;
    })
    .each(function( index, element ) {
      var additionalOptions = $(element).siblings(".note").find('.valid');
      var original = element.onblur;
      element.onblur = null;
      $(element).on( "blur", function( evt ) {
        var val = this.value;
        var matches = additionalOptions.filter(function( index, element ) {
          return $(element).text() == val;
        });
        if ( matches.length == 0 ) {
          $(this).trigger( "original", original );
        }
        else{
          element.style.backgroundColor="white";
        }
      }).on( "original", original );
    });

});


js/addTagToOnlineDesignerActionTagList.js
$(document).ready(function() {
    // Checking if field annotation is present on this page.
    if ($('#div_field_annotation').length === 0) {
        return false;
    }

    $('body').on('dialogopen', function(event, ui) {
        var $popup = $(event.target);
        if ($popup.prop('id') !== 'action_tag_explain_popup') {
            // That's not the popup we are looking for...
            return false;
        }

        // Aux function that checks if text matches the "@DEFAULT" string.
        var isDefaultLabelColumn = function() {
            return $(this).text() === '@DEFAULT';
        }

        // Getting @DEFAULT row from action tags help table.
        var $default_action_tag = $popup.find('td').filter(isDefaultLabelColumn).parent();
        if ($default_action_tag.length !== 1) {
            return false;
        }

        var tag_name = '@EXTRA-VALID-RANGES';
        var descr = 'Adds valid options outside of the min-max range for a form field with number validation. For example, by adding the action tag @EXTRA-VALID-RANGES = "88,95-97,13,15-17,24", the user will be allowed to enter any of the following numbers (in addition to the range already defined): 88, 24, any number between 15 and 17, and any number between 95 and 97.';

        // Creating a new action tag row.
        var $new_action_tag = $default_action_tag.clone();
        var $cols = $new_action_tag.children('td');
        var $button = $cols.find('button');

        // Column 1: updating button behavior.
        $button.attr('onclick', $button.attr('onclick').replace('@DEFAULT', tag_name));

        // Columns 2: updating action tag label.
        $cols.filter(isDefaultLabelColumn).text(tag_name);

        // Column 3: updating action tag description.
        $cols.last().text(descr);

        // Placing new action tag.
        $new_action_tag.insertAfter($default_action_tag);
    });
});
